Sources/Common/Communication/EventBusHandler.swift

```diff
@@ -0,0 +1,113 @@
+    ///   - eventBus: An instance of EventBus to handle event posting and observer management.
+    ///   - eventStorage: An instance of EventStorage to manage event persistence.
+    ///   - logger: A logger for logging information and errors.
+    /// Automatically loads events from file-based storage into in-memory storage upon initialization.
+    public init(eventBus: EventBus, eventStorage: EventStorage, logger: Logger) {
+        self.eventBus = eventBus
+        self.eventStorage = eventStorage
+        self.logger = logger
+        Task { await loadEventsFromStorage() }
+    }
+
+    /// Loads events from persistent storage into in-memory storage for quick access and event replay.
+    private func loadEventsFromStorage() async {
+        for eventType in EventTypesRegistry.allEventTypes() {
+            do {
+                let key = eventType.key
+                let events: [AnyEventRepresentable] = try await eventStorage.loadEvents(ofType: key)
+                await memoryStorage.storeEvents(events, forKey: key)
+            } catch {
+                logger.debug("Error loading events for \(eventType): \(error)")
+            }
+        }
+    }
+
+    /// Adds an observer for a specific event type and replays any stored events of that type to the new observer.
+    /// - Parameters:
+    ///   - eventType: The event type to observe.
+    ///   - action: The action to execute when the event is observed.
+    public func addObserver<E: EventRepresentable>(_ eventType: E.Type, action: @escaping (E) -> Void) {
+        logger.debug("EventBusHandler: Adding observer for event type - \(eventType)")
+
+        let adaptedAction: (AnyEventRepresentable) -> Void = { event in
+            if let specificEvent = event as? E {
+                action(specificEvent)
+            } else {
+                self.logger.debug("Error: Event type did not match")
+            }
+        }
+
+        Task {
+            await eventBus.addObserver(eventType.key, action: adaptedAction)
+            await replayEvents(forType: eventType)
+        }
+    }
+
+    /// Removes an observer for a specific event type.
+    /// - Parameter eventType: The event type for which to remove the observer.
+    public func removeObserver<E: EventRepresentable>(for eventType: E.Type) {
+        Task { await eventBus.removeObserver(for: E.key) }
+    }
+
+    /// Replays events of a specific type to any new observers, ensuring they receive past events.
+    /// - Parameter eventType: The event type for which to replay events.
+    private func replayEvents<E: EventRepresentable>(forType eventType: E.Type) async {
+        let key = eventType.key
+        let storedEvents = await memoryStorage.eventsForKey(key)
+
+        for event in storedEvents {
+            if let specificEvent = event as? E {
+                logger.debug("EventBusHandler: Replaying event type - \(specificEvent)")
+                let isSent = await eventBus.post(specificEvent)
```

Comment:
how do we ensure the event is sent only to the new observers?
I mean if there is an already existing observer, and then another one gets added, it seems we will call them both. Which will be duplicate call for the existing observer

Reply:
This is a good question, i do have the implementation for it but didn't add it in the first iteration to keep the implementation easier to understand. Its will be added in later PR :D